### ISSUE TITLE
[Bug] "An error occurred loading this page's template" since v7

### ISSUE THREAD

**Issue by Vinnl** [react-static/react-static#1203]
### Reporting a bug?

  Provide as much information as you can to reproduce the issue.
  **If the issue is not reproducible, it can't be fixed!**

### Environment

Run and copy the result of:

`npx envinfo --system --npmPackages react* --binaries --npmGlobalPackages react* --browsers` 

here:

```

  System:
    OS: Linux 4.15 Ubuntu 18.04.2 LTS (Bionic Beaver)
    CPU: (4) x64 Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz
    Memory: 177.36 MB / 7.52 GB
    Container: Yes
    Shell: 5.4.2 - /bin/zsh
  Binaries:
    Node: 10.15.3 - ~/.nvm/versions/node/v10.15.3/bin/node
    Yarn: 1.16.0 - /usr/bin/yarn
    npm: 6.4.1 - ~/.nvm/versions/node/v10.15.3/bin/npm
  Browsers:
    Chrome: 74.0.3729.169
    Firefox: 67.0
  npmPackages:
    react: ^16.8.6 => 16.8.6 
    react-dom: ^16.8.6 => 16.8.6 
    react-ga-donottrack: ^1.0.0 => 1.0.0 
    react-hot-loader: ^4.8.4 => 4.8.4 
    react-markdown: ^4.0.8 => 4.0.8 
    react-static: ^7.0.10 => 7.0.10 
    react-static-plugin-sass: ^7.0.10 => 7.0.10 
    react-static-plugin-typescript: ^7.0.10 => 7.0.10 
    react-syntax-highlighter: ^10.2.1 => 10.2.1 

```

of course include any other package versions here if relevant. 

### Steps to Reproduce the problem

**Base your steps off of any freshly installed react-static template!**

It's a somewhat difficult problem because it doesn't appear to be completely reproducible every time. These steps result in the problem some of the time. I have the feeling that it mostly occurs when the site is freshly loaded.

1.  Visit https://vincenttunru.com
2.  Click on one of the posts, e.g. "Fearless deployments"
3.  Hopefully see "An error occurred loading this page's template. More information is available in the console." The console then shows: "Error: export not found" in `vendors~main`.

![image](https://user-images.githubusercontent.com/4251/58384345-06cd5580-7fe1-11e9-9d70-58cd36412a97.png)

When you then navigate to a different page, the page goes completely black with the message:

> An internal error occured!
> 
> Please see the console for more details.

And the console says:

> Invariant Violation: "Minified React error #130; visit https://reactjs.org/docs/error-decoder.html?invariant=130&args[]=undefined&args[]= for the full message or use the non-minified dev environment for full errors and additional helpful warnings. 

![image](https://user-images.githubusercontent.com/4251/58406155-4475c100-8069-11e9-8bb7-6a7de18f042f.png)

It appears to be easier to reproduce when the site has not been visited for a couple of minutes...

This occurs in both Firefox and Chrome. I think it does not occur when the console is open. It also does not occur when you visit a page directly. Any pointers to other things I should investigate would be welcome.

### Expected Behavior

The expected page should just be shown every time I click a link there.

### Reproducible Demo

Source code of the website is here: https://gitlab.com/vincenttunru/blog

**Comment by sw-yx** [comment]
super weird set of symptoms. it sounds like webpack is screwing up badly some how. you may also have service worker/caching issues so be sure to check in incognito mode. no time to look at your code  for the time being

**Comment by Vinnl** [comment]
I already checked in Incognito mode, as well as in Tor browser, but it also happens there. Does React Static set up a service worker by default?

**Comment by sw-yx** [comment]
no :(

**Comment by Pixelatex** [comment]
Taking a  quick look in your code here... need that familiar with TS but it seems like you aren't returning a component for your posts...

The react error also confirms that something is not returning a valid react element.

```javascript
  const postRoutes = aliases.map((alias) => ({
            path: alias,
            redirect: postFilenameToPath(post.filename),
          }));

          routes = routes.concat(postRoutes);
```
You are only returning a path and a redirect value for your `postRoutes`.

**Comment by Vinnl** [comment]
@Pixelatex Is it required to return a component when doing a redirect? And if so, what should that component render?

**Comment by RohitRajendran** [comment]
@Vinnl Did you find a fix for this? We're also seeing the same error popup

**Comment by Vinnl** [comment]
@RohitRajendran I didn't, unfortunately :( I can still reproduce it pretty often, but haven't been able to pinpoint the cause: as far as I can see there's no invariant that's always the case when the bug occurs, and not when it doesn't. I was thinking it might be a caching issue, but any cache that might have been relevant will have been invalidated many times over by now...

If you do find a fix, please do report back :)

**Comment by t3rmian** [comment]
I've been encountering the same issue, with and without service worker, since I started using v7, though I thought it was only specific to me. So, recently I've started looking for the source; found node plugin API and a way to [reconfigure the webpack](https://github.com/react-static/react-static/blob/b26543fea37a8a3840b0ad9daa45e9a583fdd4df/docs/plugins/node-api.md#webpack). I managed to pinpoint the issue to [splitChunks](https://webpack.js.org/plugins/split-chunks-plugin/). I added *node_modules* chunks set as 'initial'. It might not be the best solution but the error is not appearing anymore [at least in my case](https://github.com/t3rmian/react-static-teapot/commit/cae585f1169a2863a8607fe3f148031dded33f71).

**Comment by SleeplessByte** [comment]
Wow. That is great @t3rmian . I can not figure out myself how or what this really causes internally. @Vinnl can you check if this works if you reconfigure webpack as described by @t3rmian ?

**Comment by Vinnl** [comment]
@SleeplessByte I've [given it a try](https://gitlab.com/vincenttunru/blog/commit/9af8cd785bb780f5d720d41e73169d4ed5928c90), but -though it might be because it's still cached somewhere- I still occasionally encounter it when clicking around [on my blog](https://vincenttunru.com/).

Does anyone know of a way to detect this error from happening? Would be good if I could log occurrences to my analytics, to see whether I see them occur less often.

**Comment by Vinnl** [comment]
I've tried it again just now, and I can no longer reproduce it. So it does seem like the issue is somewhere in the `splitChunks` setup.

I do now sometimes get a "Loading..." page, in about the same situations as where I would get the error earlier.

**Comment by Vinnl** [comment]
_Sigh..._ This just happened:

![image](https://user-images.githubusercontent.com/4251/64318288-036f4000-cfba-11e9-9088-cbd317d8c39a.png)

**Comment by SleeplessByte** [comment]
Fun story @Vinnl , the only moment I hit that on your site is if I _refresh via F5_. A hard refresh solves it...

**Comment by nsf** [comment]
This is the bug I believe, had the same issue spent a few hours debugging:
https://github.com/faceyspacey/react-universal-component/issues/198

**Comment by nsf** [comment]
Ok, here's how to "fix" it. Although be aware, it's a horrible hack.

Create a local plugin:
```
plugins/react-static-plugin-fix-universal-component/
├── node.api.js
└── react-universal-component
    ├── .babelrc.js
    ├── flowTypes.js
    ├── helpers.js
    ├── index.js
    ├── report-chunks.js
    ├── requireById
    │   └── index.js
    ├── requireUniversalModule.js
    └── utils.js
```

Plugin contains a copy of .js files from node_modules/react-universal-component/dist (very bad).

You need to edit `plugins/react-static-plugin-fix-universal-component/react-universal-component/index.js` and comment out this section in "preload" static function:
```
        // try {
        //   mod = requireSync(props, context);
        // } catch (error) {
        //   return Promise.reject(error);
        // }
```

Then node.api.js contents:
```
import path from "path";

process.env.REACT_STATIC_UNIVERSAL_PATH = path.join(__dirname, "react-universal-component/index.js");

export default () => ({
  webpack: (config, { stage }) => {
    if (stage === "prod") {
      config.resolve.alias["react-universal-component"] = path.join(__dirname, "react-universal-component/index.js");
    }
    return config;
  },
});
```

Then .babelrc.js:
```
module.exports = {
  sourceType: "unambiguous",
  presets: ["react-static/babel-preset.js"],
  plugins: [
    [
      require.resolve("@babel/plugin-transform-runtime"),
      {
        corejs: false,
        useESModules: true,
      },
    ],
    require.resolve("@babel/plugin-syntax-dynamic-import"),
  ],
};
```

While it all is a horrible hack, but that's the essence of the problem there.

**Comment by SleeplessByte** [comment]
I think this might be it @nsf. Amazing, thank you. I've also narrowed it down to universal component, but no clue on what actually broke there.

We'll work on getting a monkeypatch in until it's fixed upstream (or maybe forego the dep alltogether).

**Comment by Vinnl** [comment]
Wow @nsf, just managed to read through your analysis, and that's some excellent detective work you did there! Very much looking forward to not running into this bug any more.

Sorry for spamming everyone, but I really wanted to voice my appreciation here :)

**Comment by nsf** [comment]
I want to add to this issue: there is another possible problem going on.

You might want to comment out try/catch blocks with requireSync in "init" and "componentDidUpdate" functions as well. It's not confirmed and I didn't do a lot of research on this. But here's the theory:

0.To refresh the memory: requireSync in general is just broken, as we know on async require webpack loads the package itself and its deps in parallel, checking if package is loaded doesn't guarantee that it's ready to be initialized. Webpack puts package object into its global table as soon as <script> is loaded, but its deps might still be in progress.
1. Let's say you have a website with an index page, which has some links to some other pages in a header.
2. You load the index page, react-static preloading kicks in.
3. Without waiting you click one of the links.
4. Now you get to a page with another react universal component and as such those "init" and potentially due to some reason "componentDidUpdate" functions are called which do use "requireSync".
5. And since preloading is in progress at that moment this pretty much leads to the same bug as described in this issue.

Just another word of warning here. What react-universal-component does in "requireSync" is not correct. It makes assumptions about webpack that are not right. Please don't make assumptions about webpack in general. If you do async require, then do async require always, don't go to `__webpack_modules__` and attempt to initialize/use the module yourself. Not to blame anyone, we all learn how to use async code properly in our codebases, I know it's complicated and often it seems like having a synchronous path is a good and easy to reason about thing. But more often than not you just have to embrace the async.

**Comment by SleeplessByte** [comment]
> But more often than not you just have to embrace the async.

Very true. And it's possible to make any sync code async, but not the other way around. The pre-loading issue here _could_ be the case, but I don't know if it is. Preloading links actually preloads the route data, not the actual components on the page. In other words: it's not webpack loading, and pre RS 8, there is only one chunk by default IIRC...

**Comment by tomduncalf** [comment]
Is this still an issue? I've had a couple of instances of this error on a new react-static site reported. I've not actually seen it happen myself. Worrying if there's a random risk of errors occuring!

**Comment by Vinnl** [comment]
@tomduncalf Yes, this hasn't been fixed yet as far as I'm aware. Note that I think I'm able to consistently reproduce it now. Everything works fine if you click a link quickly after page load, but if you wait to long, clicking a link that's dynamically loaded will make this error show up.

(Incidentally, just this weekend I migrated my blog to Next.js which now also supports static site generation, so the steps for reproduction mentioned in the original comment no longer apply.)

**Comment by tomduncalf** [comment]
@Vinnl, thanks for the update. That's a shame to hear :(

Slight off topic, how was the migration to Next.js? I've had quite a few random issues with this version of react-static (that I didn't have with v6) that makes me wonder if I should just bite the bullet and migrate to something else – issues like this which can break the site and are not resolved are especially concerning (no offence intended to the maintainers).

**Comment by nsf** [comment]
@tomduncalf As I was also having issues with this, I could answer your question.

I migrated all of the websites I had to work on (three of them) to gatsby. Gatsby is quite similar to react-static, there's also a data model and a config thing where you can generate pages and all that. As for react-static, I've spent a lot of time trying to fix all the bugs myself with ugly and horrible hacks involving https://www.npmjs.com/package/patch-package, but at the end I never quite reached the point where it was behaving well. Some strange loading artifacts kept popping up, so I ended up with moving to gatsby. It worked out well. There's one issue with gatsby though is that unlike react-static you can't explicitly load page's data, and in case if you need that feature, it can be done via hacks, e.g.:
```typescript
export async function loadPage<T>(path: string): Promise<T> {
  const data = await (window as any).___loader.loadPage(path);
  return data.json.pageContext;
}
```

It's sad that we had to move, but at the end of day we pick what works best.

**Comment by tomduncalf** [comment]
Thanks @nsf, Gatsby sounds promising and was one of the solutions I originally considered when doing v1 of this site, but I chose react-static as it seemed a lot simpler. 

Shame as I had a good experience with v6, and mostly with v7 (the main bug I've found is needing to restart the dev server whenever any significant changes are made to a page), and I'd have to do a bunch of unpaid work to migrate it... but it might be worthwhile in the long run. I'll see if I get any more reports of this error first!

**Comment by Vinnl** [comment]
@tomduncalf Next.js actually feels quite similar to React Static now (at least the static site generation part - obviously it can do a lot more), but of course, you're still changing the base out from under your application. It took me quite a bit of time, though large parts of that were just learning how Next.js does things.

**Comment by SleeplessByte** [comment]
> but it might be worthwhile in the long run if maintenance of this package is an issue. 

COVID-19 has bascially hit all the _paid customers_ that were using react-static as underlying software, which is why the development has been grinding to a halt. There's nothing really we can do about that at this time, other than relying on contributors to help out.

`next.js` and `gatsby` are amazing platforms, and if they suit your needs, you should definitely try them out. We have some of our clients running on those as well.

`react-static` is definitely simpler in terms of complexity.

**Comment by Vinnl** [comment]
Just to emphasise @SleeplessByte, I'm very appreciative of you and your company stepping in to take over maintenance when that was most needed :) It's just that, given the circumstances, migration to Next.js made sense; hope that didn't come across as a dig at this project/your efforts, as that wasn't the intention.

**Comment by SleeplessByte** [comment]
Of course not! We actively support migrations from and to other platforms. We should all use what works for us best and we appreciate the continued contributions from everyone.

**Comment by tomduncalf** [comment]
Yeah, likewise (what's the story with the library being taken over btw, is there a link about that?) - totally appreciate your hard work and I know it's a hard time right now! Hopefully I can fix this error and keep on using react-static

**Comment by SleeplessByte** [comment]
@tomduncalf  you'll want to check #1262 . 

Long story short, Tanner was working on this paid, but no longer needed it for his clients. We literally were in the same position, so made some strides in the beginning, but now we're also out of clients. We haven't discussed dropping support altogether, because there's a good chance the clients will come back when COVID-19 dies down, but we are looking for co-maintainers.

**Comment by tomduncalf** [comment]
Got ya - I hope the work comes back soon!

**Comment by tomduncalf** [comment]
Hey @Vinnl, just wanted to say thank you for mentioning your migration to Gatsby. After hitting this (or a similar) issue for myself today (previously it had just been random user reports) and spending a bit of time trying to work out what was going on (non-starter - one of my browsers just seemed to be stuck with a "broken" version of the site which would randomly 404), I decided to try the migration and was actually amazed how easy it was, it took me about 90 minutes from first visiting the Gatsby website to having the whole site (only 6 templates) migrated!

You can actually use Gatsby with a very similar model to the `static.config.js` way if you don't want to mess with GraphQL etc.

For anyone else who wants to try this, at a high level the changes I had to make were:
* Install `gatsby` from npm
* Create a `gatsby-node.js` file to fetch data and create routes using `createPage` as described in https://www.gatsbyjs.org/docs/using-gatsby-without-graphql/ - basically a straight migration of `static-config.js`. This means you aren't using all the GraphQL stuff, but seemed the quickest way to migrate
* Converted the site from having a top-level `<App/>` component with the global layout and then a `<Router/>` to one "template" file per page using a common `<Layout/>` component in there for global layout (was fairly easy as I already had "containers" for each page)
* Converted the site from `useSiteData`/`useRouteData` to getting the data at the "template" level from props and passing it down (I prefer the way react-static does this, feels neater!)
* Swap out `reach-router` `<Link/>` for the Gatsby implementation
* Add plugins e.g. for styled components, Google Analytics

I can try to share more detail if anyone is interested.

It's a shame to move away from react-static as I do prefer its API and simplicity, but this issue (and a couple of other random ones I had e.g. often adding a new component when editing a page required a restart of the dev server due to some "fibre" error) made it necessary as I can't risk having client sites randomly erroring. The larger ecosystem and more "batteries included" approach of Gatsby is also a plus e.g. installing GA with a simple plugin, no need for hacks to ensure the page scrolls to the top correctly on navigation etc, and it seems the builds are faster.

Thanks to the react-static mantainers for all their work on this project, I don't mean to be negative by posting this but just sharing my experiences for anyone else in a similar situation as it sounds like maintenance on react-static may be somewhat sporadic for the forseeable future.

Cheers, all the best,
Tom

**Comment by RossRogers** [comment]
I've been loving react-static.  I'd hate to see this bug be the bathwater that causes the baby to be thrown out.  I can confirm that @nsf 's hack seems to work.

**Comment by tomduncalf** [comment]
Glad the fix works for you @RossRogers, couldn’t get it to work for me. I’ve since ported another site to Gatsby but then to NextJS because Next has an awesome preview mode which will reload the API calls on every refresh - makes life so much easier when editing. 

Based on my experience I’d say Gatsby is easier to move to from React Static, but NextJS has some nicer features. Was definitely more work though, but unless you go all in with the GraphQL stuff on Gatsby you don’t get an equivalent refresh feature as far as I could see (and retrofitting the GraphQL would be more work than moving to NextJS, I reckon). 

So hopefully it doesn’t come to this for you, but if it does I’d say both are good options but NextJS is worth the extra bit of effort required to move to I think.

**Comment by RossRogers** [comment]
Yeah, I've tried Gatsby and I moved to react-static from that.  GraphQL is SOAP 2.0, IMHO.  NextJS debug messages were unusably atrocious when I tried to move.  React-static is simpler and I love it [for now].

**Comment by tomduncalf** [comment]
@RossRogers I was using Gatsby without the GraphQL stuff, you can do that using their `createPage` API... but you do miss out on some stuff, in particular the ability to "live reload" a page's content. Agreed with you about GraphQL. NextJS wasn't too painful, I guess they've improved, but I agree that I liked react-static's simplicity and I'm sad that this bug pushed me away from it.

**Comment by nickjohngray** [comment]
> Ok, here's how to "fix" it. Although be aware, it's a horrible hack.
> 
> Create a local plugin:
> 
> ```
> plugins/react-static-plugin-fix-universal-component/
> ├── node.api.js
> └── react-universal-component
>     ├── .babelrc.js
>     ├── flowTypes.js
>     ├── helpers.js
>     ├── index.js
>     ├── report-chunks.js
>     ├── requireById
>     │   └── index.js
>     ├── requireUniversalModule.js
>     └── utils.js
> ```
> 
> Plugin contains a copy of .js files from node_modules/react-universal-component/dist (very bad).
> 
> You need to edit `plugins/react-static-plugin-fix-universal-component/react-universal-component/index.js` and comment out this section in "preload" static function:
> 
> ```
>         // try {
>         //   mod = requireSync(props, context);
>         // } catch (error) {
>         //   return Promise.reject(error);
>         // }
> ```
> 
> Then node.api.js contents:
> 
> ```
> import path from "path";
> 
> process.env.REACT_STATIC_UNIVERSAL_PATH = path.join(__dirname, "react-universal-component/index.js");
> 
> export default () => ({
>   webpack: (config, { stage }) => {
>     if (stage === "prod") {
>       config.resolve.alias["react-universal-component"] = path.join(__dirname, "react-universal-component/index.js");
>     }
>     return config;
>   },
> });
> ```
> 
> Then .babelrc.js:
> 
> ```
> module.exports = {
>   sourceType: "unambiguous",
>   presets: ["react-static/babel-preset.js"],
>   plugins: [
>     [
>       require.resolve("@babel/plugin-transform-runtime"),
>       {
>         corejs: false,
>         useESModules: true,
>       },
>     ],
>     require.resolve("@babel/plugin-syntax-dynamic-import"),
>   ],
> };
> ```
> 
> While it all is a horrible hack, but that's the essence of the problem there.

This is grate, it worked for me! i had to provide the plugin path, in static.config and not use the plugins folder.
I made a plugin out of your code steps:  zip file attached .
This error was happing to me  often, i had to double or triple  refresh to get rid of it, after a new deployment on netify.

[react-static-plugin-fix-universal-component.zip](https://github.com/react-static/react-static/files/5222007/react-static-plugin-fix-universal-component.zip)

**Comment by brokerslurp** [comment]
I'm seeing this issue pop up now on my configuration of react-static.  It seems to randomly happen (can't consistently reproduce).  Are people still using the above hack to fix?  What is the state of this issue?

**Comment by SleeplessByte** [comment]
People still use the above hack. It's an upstream issue in react-universal-component.

**Comment by brokerslurp** [comment]
> People still use the above hack. It's an upstream issue in react-universal-component.

Ok, thanks for the info.

Is there an alternative to using react-universal-component?  If the upstream dependency is not going to fix the bug, then maybe react-static shouldn't continue to use it..
